### PR TITLE
[VPlan] Generalize collectUsersInExitBlocks for multiple exit bbs.

### PR DESCRIPTION
**Pull request by fhahn**
Generalize collectUsersInExitBlock to collecting exit users in multiple exit blocks. Exit blocks are leaf nodes in the VPlan (without successors) except the scalar header.

Split off in preparation for https://github.com/llvm/llvm-project/pull/112138

### CODE REVIEW

**Comment by fhahn** [comment]
> Hi @fhahn, it doesn't look like #112138 has to deal with users of exit values, which this patch says it's in preparation for. Do you plan to expand the scope of PR #112138 to include more of PR #88385?
> 
> I am very grateful for your help and feedback in trying to drive PR #88385 forward, but I was under the impression that PR #112138 was simply presenting an alternative basic skeleton for handling multiple loop region successors in VPlan, from which I could then continue to drive PR #88385 forwards.
> 

This is trying to avoid having to add extra code dealing with the exit phis following on the suggestions/comments in https://github.com/llvm/llvm-project/pull/112138/files#r1810491008.  https://github.com/llvm/llvm-project/pull/112138 still only handles live-ins in exit phis, but in the latest version it doesn't have to update the VPIRInstructions; instead it rejects any VPlans that contains VPIRInstructions without operands (i.e. because they are live-out inductions) or any of their added operands aren't live-ins. Alternatively I could also restore the code from https://github.com/llvm/llvm-project/pull/112138/files#r1810491008, WDYT?

**Comment by fhahn** [comment]
> This is trying to avoid having to add extra code dealing with the exit phis following on the suggestions/comments in https://github.com/llvm/llvm-project/pull/112138/files#r1810491008. #112138 still only handles live-ins in exit phis, but in the latest version it doesn't have to update the VPIRInstructions; instead it rejects any VPlans that contains VPIRInstructions without operands (i.e. because they are live-out inductions) or any of their added operands aren't live-ins. Alternatively I could also restore the code from https://github.com/llvm/llvm-project/pull/112138/files#r1810491008, WDYT?

IIRC we briefly talked about having support for phis that use only live-outs to test the whole structure more end-to-end, but another alternative would be to check the exit blocks for any VPIRInstruction instead of the current check around https://github.com/llvm/llvm-project/pull/112138/files#diff-da321d454a7246f8ae276bf1db2782bf26b5210b8133cb59e4d7fd45d0905decR9196

**Comment by david-arm** [comment]
> > This is trying to avoid having to add extra code dealing with the exit phis following on the suggestions/comments in https://github.com/llvm/llvm-project/pull/112138/files#r1810491008. #112138 still only handles live-ins in exit phis, but in the latest version it doesn't have to update the VPIRInstructions; instead it rejects any VPlans that contains VPIRInstructions without operands (i.e. because they are live-out inductions) or any of their added operands aren't live-ins. Alternatively I could also restore the code from https://github.com/llvm/llvm-project/pull/112138/files#r1810491008, WDYT?
> 
> IIRC we briefly talked about having support for phis that use only live-outs to test the whole structure more end-to-end, but another alternative would be to check the exit blocks for any VPIRInstruction instead of the current check around https://github.com/llvm/llvm-project/pull/112138/files#diff-da321d454a7246f8ae276bf1db2782bf26b5210b8133cb59e4d7fd45d0905decR9196

Hi @fhahn, thanks for the quick response! When I was looking at this patch yesterday PR #112138 hadn't been updated - I think there must have been an incorrect version pushed up because you'd addressed some comments but the code hadn't changed. So I was lacking some context about what how this patch fits in with that, but now I see your updated version and it makes more sense now! And I understand you're trying to limit the patch to only handling live-ins, which is fine. I'm currently reviewing your latest version of PR #112138 first, then I'll come back to this.

File: llvm/lib/Transforms/Vectorize/VPlanCFG.h
Comment (david-arm) [inline llvm/lib/Transforms/Vectorize/VPlanCFG.h:315]: Same comment as in the other PR - I think this can return different blocks depending upon whether you call this before or during VPlan::execute. Is this something we should worry about?

File: llvm/lib/Transforms/Vectorize/VPlan.h
Comment (david-arm) [inline llvm/lib/Transforms/Vectorize/VPlan.h:3842]: If possible, it would be nice if this returned a concrete type or had a comment describing it. It's just a bit awkward for the reader having to look in a different header file to see what it returns.

File: llvm/lib/Transforms/Vectorize/LoopVectorize.cpp
Comment (david-arm) [inline llvm/lib/Transforms/Vectorize/LoopVectorize.cpp:8840]: Not sure how this works when all exiting blocks jump to the same exit block? LoopVectorizationLegality has a list of exiting and exit blocks that could perhaps be useful here? Otherwise you could compile a list of all exiting blocks for that exit block and walk over them.

This algorithm is definitely more powerful than what I originally planned in https://github.com/llvm/llvm-project/pull/88385, which isn't a bad thing. However, I guess it is more difficult to implement for the general case. If it's easier you could just limit this to the single early exit case that LoopVectorizationLegality current supports? That class already has the single early exiting block and early exit block.

**Review by david-arm — COMMENTED** [review]
Thanks for this patch @fhahn. I just had a few comments!

**Review by fhahn — COMMENTED** [review]
Also added an assert to make sure we don't process any exit values we cannot handled yet (non-live ins from blocks other than MiddleVPBB)

File: llvm/lib/Transforms/Vectorize/LoopVectorize.cpp
Comment (david-arm) [reply llvm/lib/Transforms/Vectorize/LoopVectorize.cpp:8841]: Oh I see so this patch is only generalising for multiple exit blocks on the assumption that there is only a single predecessor per exit block. It looks like in PR #112138 you've generalised this further by removing that restriction. I guess this is ok and reduces the diff for PR #112138. In that case is it worth adding a temporary assert until PR #112138 lands?

```
  assert(ExitingBB && "Expected exit block to have a single predecessor");
```

File: llvm/lib/Transforms/Vectorize/VPlanCFG.h
Comment (david-arm) [reply llvm/lib/Transforms/Vectorize/VPlanCFG.h:315]: Ah, I think the reason why this works is because of the extra `VPIRBB->getNumSuccessors() == 0` condition, since the middle block will always have at least one successor.

Comment (fhahn) [reply llvm/lib/Transforms/Vectorize/VPlanCFG.h:315]: The middle-block may or may not be included by `VPBlockUtils::blocksOnly<VPIRBasicBlock>`, but the middle block will never be considered as an exit block, as it has successors.

Comment (david-arm) [reply llvm/lib/Transforms/Vectorize/VPlanCFG.h:315]: Yeah, thanks for explaining! I realised that around the same time that you replied. It just took me a while to figure it out. :)

**Review by david-arm — APPROVED** [review]
LGTM! I had a minor comment about potentially adding an assert that `ExitingBB` is non-null - I'll leave that to you if you think it's worth it or not.

File: llvm/lib/Transforms/Vectorize/LoopVectorize.cpp
Comment (fhahn) [reply llvm/lib/Transforms/Vectorize/LoopVectorize.cpp:8841]: The current version of the patch has an assert in `addUsersInExitBlocks` to make sure we only updated the exit phis, if the incoming value is from the middle block. Do you think that's sufficient or should I add another one here as suggested? 

I'll update #112138 once this lands, and the assert in  `addUsersInExitBlocks` would be turned into a bail-out as it means we encountered a exit value we cannot handle yet, and only will be able to with #88385,?

Comment (david-arm) [reply llvm/lib/Transforms/Vectorize/LoopVectorize.cpp:8841]: Ah yes I see, that makes sense. In that case you're right what you have now is fine and happy for you to merge that patch!

**Pull request merged**
